File: backend/routers/exercises.py
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, ConfigDict
from sqlalchemy.orm import Session

from auth import get_current_user
from database import get_db
from models.exercise import Exercise

router = APIRouter(prefix="/api/exercises", tags=["exercises"])

SESSIONS = ["Push A", "Pull A", "Legs A"]


class ExerciseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    muscles: str | None
    session: str
    position: int
    type: str
    equip: str


@router.get("/sessions")
def list_sessions(_: str = Depends(get_current_user)) -> list[str]:
    return SESSIONS


@router.get("")
def list_exercises(
    session: str,
    db: Session = Depends(get_db),
    _: str = Depends(get_current_user),
) -> list[ExerciseOut]:
    if session not in SESSIONS:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Unknown session")
    exercises = (
        db.query(Exercise)
        .filter(Exercise.session == session)
        .order_by(Exercise.position)
        .all()
    )
    return exercises

File: frontend/src/api/exercises.ts
import { request } from './client'

export interface Exercise {
  id: number
  name: string
  muscles: string | null
  session: string
  position: number
  type: string
  equip: string
}

export function getSessions(): Promise<string[]> {
  return request<string[]>('/api/exercises/sessions')
}

export function getExercises(session: string): Promise<Exercise[]> {
  return request<Exercise[]>(`/api/exercises?session=${encodeURIComponent(session)}`)
}

export function getAllExercises(session?: string, search?: string): Promise<Exercise[]> {
  const params = new URLSearchParams()
  if (session) params.set('session', session)
  if (search) params.set('search', search)
  const qs = params.toString()
  return request<Exercise[]>(`/api/exercises${qs ? `?${qs}` : ''}`)
}
